### PR TITLE
OCPBUGS-16307: default instance type for Azure based on the architecture (+ gcp)

### PR DESCRIPTION
**Pull request by aleskandro**
This PR proposes a fix for OCPBUGS-16307. It proposes to use a map to consume the default instance types based on the cloud provider and the architecture of the control plane. Therefore it refactors the previous work on AWS to use the map and adds the defaults for Azure.

Since we need similar work for GCP in the efforts for ARMOCP-15, I also added the defaults for it. If requested (for example, to ease the backport of the Azure defaults fix), I can split this PR in two (AWS refactor + Azure, and GCP).

The unit tests are updated according to the new code. It's worth noting that they were bound to the architecture of the host running them, such that we cannot test defaults for different architectures from the same testrun. Just pointing out for now, as that might be a future work to consider.

Finally, in the case of Azure, for arm64, we add the `-gen2` postfix to the gallery name. In fact, since the arm64 enablement, we create an image gallery and images for the Gen2 HyperV. It is done for both x86_64 and aarch64 by the installer. However, for x86_64 clusters installed before that change, we need yet to maintain the non -gen2 image gallery name as that might be the only one available.  In the case of multi-arch compute nodes, it is still the responsibility of the user to create the additional architecture image gallery, and the defaults are held bound to the control plane architecture.

cc @jeffdyoung @r4f4

### CODE REVIEW

File: pkg/webhooks/machine_webhook.go [hunk 1]
```diff
@@ -103,9 +111,28 @@ var (
...
+			AMD64: defaultGCPX86MachineType,
+			ARM64: defaultGCPARMMachineType,
+		},
+	}
 )
 
 const (
+	arch              = machineArch(goruntime.GOARCH)
+	ARM64 machineArch = "arm64"
+	AMD64 machineArch = "amd64"
```
Comment (JoelSpeed) [inline pkg/webhooks/machine_webhook.go:134]: How will this work when we have heterogeneous clusters?

What happens if the arch is not one of these two? I think we need to have a fallback because the tests running on a macbook will not pass given this, they will find an empty value from the map lookup

Comment (aleskandro) [reply pkg/webhooks/machine_webhook.go:119]: Hi, I changed the code to use a function; it returns the amd64 default instance type for the given provider if failing the lookup. It still returns an empty string if the function is used for a non-supported provider. Is that ok?

I also moved the map to be local to the function for "disallowing modifications", if that is worth it.

> How will this work when we have heterogeneous clusters?

In the case of heterogeneous clusters, we default to the instance type of the same architecture as the control plane. This is the same behavior previously implemented for AWS only. I don't see other ways to define a 'primary architecture' and its default instance type here.


Once the review is done and we agree on an approach, I'll squash the commits properly.

Comment (aleskandro) [reply pkg/webhooks/machine_webhook.go:119]: Rebased and added logs and warnings

**Comment by aleskandro** [comment]
/hold

gcp uses to default the diskImage as Azure.

What do we do for AWS?

File: pkg/webhooks/machine_webhook.go [hunk 2]
```diff
@@ -129,14 +149,16 @@ const (
...
 	defaultGCPCredentialsSecret = "gcp-cloud-credentials"
 	defaultGCPDiskSizeGb        = 128
 	defaultGCPDiskType          = "pd-standard"
-	// https://releases-art-rhcos.svc.ci.openshift.org/art/storage/releases/rhcos-4.8/48.83.202103122318-0/x86_64/meta.json
-	// https://github.com/openshift/installer/blob/796a99049d3b7489b6c08ec5bd7c7983731afbcf/data/data/rhcos.json#L90-L94
-	defaultGCPDiskImage = "projects/rhcos-cloud/global/images/rhcos-48-83-202103221318-0-gcp-x86-64"
-	defaultGCPGPUCount  = 1
+	// https://releases-rhcos-art.apps.ocp-virt.prod.psi.redhat.com/?stream=prod/streams/4.14-9.2&release=414.92.202307070025-0&arch=x86_64#414.92.202307070025-0
+	// https://github.com/openshift/installer/commit/0cec4e1403d78387729f21f04d0f764f63fc552e
+	defaultGCPX86DiskImage = "projects/rhcos-cloud/global/images/rhcos-414-92-202307070025-0-gcp-x86-64"
```
Comment (aleskandro) [inline pkg/webhooks/machine_webhook.go:159]: to double check if this bump is ok @JoelSpeed

**Review by JoelSpeed — COMMENTED** [review]
/lgtm
/approve
